indexed_gzip/zran_file_util.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * zran_file_util.c - file utilities used in zran.c to manipulate either
+ * Python file-like objects or file descriptors.
+ *
+ *
+ * Author: Paul McCarthy <pauldmccarthy@gmail.com>
+ */
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#include "zran_file_util.h"
+
+#ifdef _WIN32
+#define FSEEK _fseeki64
+#define FTELL _ftelli64
+#include "windows.h"
+#include "io.h"
+#else
+#define FSEEK fseeko
+#define FTELL ftello
+#endif
+
+/*
+ * Implements a method analogous to fread that is performed on Python file-like objects.
+ */
+size_t _fread_python(void *ptr, size_t size, size_t nmemb, PyObject *f) {
+    PyObject *data;
+    char *buf;
+    Py_ssize_t len;
+    if ((data = PyObject_CallMethod(f, "read", "(n)", size * nmemb)) == NULL)
+        goto fail;
+    if ((buf = PyBytes_AsString(data)) == NULL)
+        goto fail;
+    if ((len = PyBytes_Size(data)) == -1)
+        goto fail;
+    memmove(ptr, buf, (size_t) len);
+    Py_DECREF(data);
+    return (size_t) len / size;
+
+fail:
+    Py_XDECREF(data);
+    return 0;
+}
+
+/*
+ * Implements a method analogous to ftell that is performed on Python file-like objects.
+ */
+int64_t _ftell_python(PyObject *f) {
+    PyObject *data;
+    uint64_t result;
+    if ((data = PyObject_CallMethod(f, "tell", NULL)) == NULL)
+        goto fail;
+    if ((result = PyLong_AsUnsignedLong(data)) == (unsigned long)-1 && PyErr_Occurred())
+        goto fail;
+    Py_DECREF(data);
+    return result;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to fseek that is performed on Python file-like objects.
+ */
+int _fseek_python(PyObject *f, int64_t offset, int whence) {
+    PyObject *data;
+    if ((data = PyObject_CallMethod(f, "seek", "(l,i)", offset, whence)) == NULL)
+        goto fail;
+    Py_DECREF(data);
+    return 0;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to feof that is performed on Python file-like objects.
```

Comment:
```suggestion
 * Implements a method analogous to feof that is performed on Python file-like objects.
 * This method requires the file size to be input, because there is no other way to tell if a Python file-like is at EOF without reading it.
```